### ISSUE TITLE
How to perform ONNX model evaluation after conversion from RT-DETR

### ISSUE THREAD

**Issue by isuchy** [ultralytics/ultralytics#15733]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I was able to train RT-DETR using [Roboflow's notebook](https://github.com/roboflow/notebooks/blob/main/notebooks/train-rt-detr-on-custom-dataset-with-transformers.ipynb) on my own dataset in COCO format. I then converted the model to ONNX format, which I plan to convert to TFLite format later. For this purpose, I used the TorchScript `torch.onnx.export` ONNX exporter.

Now I need to evaluate the model in ONNX format and see if metrics like mAP50_95, mAP50 and mAP75 are comparable to the original model.

For evaluating a model for example YOLOv10, I would use the val.py script and get the given metrics that way. 
```
# Load trained model
model = YOLOv10('/content/yolov10_best.onnx')

# Evaluate the model's performance on the test set
metrics = model.val(data='/content/datasets/data.yaml')
print(metrics.box.map)  # Print mAP50-95
print(metrics.box.map50)  # Print mAP50
print(metrics.box.map75)  # Print mAP75
```
The problem is that the data argument in the val() function requires a yaml subfile, which is the format used by YOLO. The COCO format uses a different system for dataset and annotation and cannot be used in this case.

Could anyone advise how to calculate the accuracy of the ONNX model?

**Comment by glenn-jocher** [comment]
@isuchy to evaluate your ONNX model, you can use the `onnxruntime` package to run inference and then compare the results with your ground truth annotations. Unfortunately, the `val()` function in YOLOv10 expects a specific dataset format. You may need to write a custom evaluation script to handle COCO format annotations. For detailed guidance on using `onnxruntime`, please refer to the [ONNX Runtime documentation](https://onnxruntime.ai/docs/).

**Comment by isuchy** [comment]
Hi @glenn-jocher, thanks for your reply. I know what you described, that I need to use the onnxruntime package to run the inference and then compare the results with the ground truth annotations, but I don't have the experience to write my own script for that purpose. Even chatGPT didn't offer me a working solution 😅
So I decided to try to convert the COCO dataset format to YOLO format and I used your `convert_coco()` function from the `ultralytics.data.converter` package. The code looks like this.

```
from ultralytics.data.converter import convert_coco

convert_coco(
    labels_dir="/content/coco/valid",
    save_dir="coco_converted/",
    use_segments=False,
    use_keypoints=False,
    cls91to80=False,
)
```
I had to move the class index, because I have only one class in the training dataset, so after the conversion I had -1 instead of 0. Then the content of the individual txt files with bboxes in the labels directory looked OK and I set the correct path to the val attribute in the data.yaml file and inserted this modified data.yaml as a parameter to the `model.val()` function.
```
from ultralytics import YOLO

# Load trained model
model = YOLO('/content/exported_model.onnx')

# Evaluate the model's performance on the test set
metrics = model.val(data='/content/datasets/data.yaml')
print(metrics.box.map)  # Print mAP50-95
print(metrics.box.map50)  # Print mAP50
print(metrics.box.map75)  # Print mAP75
```

this is what the output from that cell looks like

```
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics YOLOv8.2.79 🚀 Python-3.10.12 torch-2.3.1+cu121 CPU (Intel Xeon 2.20GHz)
Loading /content/exported_model.onnx for ONNX Runtime inference...
WARNING ⚠️ Metadata not found for 'model=/content/exported_model.onnx'
Setting batch=1 input of shape (1, 3, 640, 640)
Downloading https://ultralytics.com/assets/Arial.ttf to '/root/.config/Ultralytics/Arial.ttf'...
100%|██████████| 755k/755k [00:00<00:00, 14.7MB/s]
val: Scanning /content/dataset/labels.cache... 146 images, 0 backgrounds, 0 corrupt: 100%|██████████| 146/146 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 146/146 [09:05<00:00,  3.74s/it]                   all        146        517          0          0          0          0

Speed: 1.8ms preprocess, 3711.2ms inference, 0.0ms loss, 0.5ms postprocess per image
Results saved to runs/detect/val5
0.0
0.0
0.0
```
Since all values are zero, I conclude that `convert_coco()` didn't work correctly and the coordinates for the bounding boxes were wrongly evaluated. 
Do you think I did the right thing and is it even possible to do it this way? Or is the only option left to me is to try to write my own script like you wrote above?

**Comment by glenn-jocher** [comment]
@isuchy it looks like you're on the right track with converting your COCO dataset to YOLO format. The zero values in your evaluation suggest there might be an issue with the conversion or the data paths. Ensure your `data.yaml` file correctly points to the converted labels and images. If the issue persists, you may need to manually verify the converted annotations. For further assistance, please refer to the [Ultralytics documentation](https://docs.ultralytics.com/integrations/onnx/).

**Comment by isuchy** [comment]
Hi again @glenn-jocher ,
So following your suggestion I tried to manually convert COCO to YOLO format and I used the following script. 

```
import json
import os

def convert_coco_to_yolo(coco_json_path, images_dir, output_dir):
    # Load COCO JSON file
    with open(coco_json_path, 'r') as f:
        coco_data = json.load(f)

    # Create output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)

    # Create a mapping from category id to class index
    category_id_to_class_index = {category['id']: idx for idx, category in enumerate(coco_data['categories'])}

    # Process each image in the COCO dataset
    for image_info in coco_data['images']:
        image_id = image_info['id']
        image_width = image_info['width']
        image_height = image_info['height']
        image_filename = image_info['file_name']

        # Prepare the YOLO annotation file
        yolo_annotation = []

        # Find all annotations for this image
        for annotation in coco_data['annotations']:
            if annotation['image_id'] == image_id:
                category_id = annotation['category_id']
                bbox = annotation['bbox']  # [x, y, width, height]

                # Convert COCO bbox to YOLO format
                x_center = (bbox[0] + bbox[2] / 2) / image_width
                y_center = (bbox[1] + bbox[3] / 2) / image_height
                width = bbox[2] / image_width
                height = bbox[3] / image_height

                # Get class index
                class_index = category_id_to_class_index[category_id]

                # Append to YOLO annotation list
                yolo_annotation.append(f"{class_index} {x_center} {y_center} {width} {height}")

        # Write YOLO annotations to file
        yolo_filename = os.path.splitext(image_filename)[0] + '.txt'
        yolo_filepath = os.path.join(output_dir, yolo_filename)
        with open(yolo_filepath, 'w') as f:
            f.write('\n'.join(yolo_annotation))

# Example usage
coco_json_path = '/content/coco/valid/_annotations.coco.json'  # Path to your COCO JSON file
images_dir = '/content/coco/valid'  # Path to your images directory
output_dir = '/content/output/yolo/val'  # Path to save YOLO annotations

convert_coco_to_yolo(coco_json_path, images_dir, output_dir)
```

Here is an example of the processing of one image with the corresponding annotation.
Snippet of COCO format
```
{
      "iscrowd": 0,
      "image_id": 303,
      "bbox": [
        974.0,
        451.0,
        101.0,
        116.0
      ],
      "segmentation": [
        [
          1032,
          463,
          1059,
          477,
          1072,
          500,
          1075,
          516,
          1074,
          534,
          1060,
          536,
          1064,
          560,
          1045,
          567,
          1032,
          564,
          1020,
          553,
          1011,
          550,
          1008,
          559,
          990,
          560,
          982,
          550,
          974,
          525,
          977,
          512,
          974,
          480,
          990,
          469,
          998,
          454,
          1014,
          451,
          1024,
          458
        ]
      ],
      "category_id": 1,
      "id": 975,
      "area": 8636
    },
    {
      "iscrowd": 0,
      "image_id": 303,
      "bbox": [
        1134.0,
        568.0,
        117.0,
        172.0
      ],
      "segmentation": [
        [
          1228,
          616,
          1244,
          603,
          1251,
          585,
          1237,
          568,
          1219,
          568,
          1197,
          584,
          1174,
          585,
          1167,
          598,
          1164,
          633,
          1165,
          646,
          1152,
          672,
          1146,
          694,
          1135,
          703,
          1134,
          717,
          1139,
          726,
          1163,
          718,
          1178,
          700,
          1182,
          725,
          1182,
          740,
          1208,
          740,
          1236,
          734,
          1218,
          720,
          1232,
          704,
          1228,
          680,
          1231,
          656,
          1235,
          640
        ]
      ],
      "category_id": 1,
      "id": 976,
      "area": 11704
    },
    {
      "iscrowd": 0,
      "image_id": 303,
      "bbox": [
        281.0,
        765.0,
        158.0,
        136.0
      ],
      "segmentation": [
        [
          363,
          807,
          397,
          818,
          414,
          794,
          430,
          799,
          433,
          823,
          439,
          844,
          430,
          865,
          402,
          870,
          380,
          882,
          378,
          896,
          338,
          901,
          311,
          883,
          295,
          858,
          281,
          835,
          289,
          810,
          332,
          781,
          343,
          765,
          360,
          767,
          373,
          774,
          374,
          787
        ]
      ],
      "category_id": 1,
      "id": 977,
      "area": 12987
    },
    {
      "iscrowd": 0,
      "image_id": 303,
      "bbox": [
        659.0,
        775.0,
        138.0,
        149.0
      ],
      "segmentation": [
        [
          748,
          784,
          731,
          777,
          707,
          776,
          693,
          789,
          670,
          823,
          659,
          858,
          676,
          870,
          697,
          881,
          704,
          894,
          697,
          916,
          707,
          924,
          728,
          924,
          749,
          918,
          743,
          903,
          763,
          892,
          772,
          869,
          774,
          849,
          782,
          822,
          797,
          796,
          780,
          777,
          755,
          775
        ]
      ],
      "category_id": 1,
      "id": 978,
      "area": 12974
    }
```

YOLO format after conversion
```
1 0.53359375 0.4678308823529412 0.05260416666666667 0.10661764705882353
1 0.62109375 0.6011029411764706 0.0609375 0.15808823529411764
1 0.1875 0.765625 0.08229166666666667 0.125
1 0.37916666666666665 0.7807904411764706 0.071875 0.13694852941176472
```

A here is my `data.yaml`
```
train: /content/yolo_dataset/train/images
val: /content/yolo_dataset/val/images  
test: /content/yolo_dataset/test/images

nc: 1
names: ["soldier"]
```

The source RT-DETR model was smoothly converted to ONNX, no errors or warnings.
Values of the original model after training on custom dataset:
map50_95: 0.42
map50: 0.66
map75: 0.46

Can you please advise me what else I should look for? I don't see any fault and I have no idea why the evaluation is not working correctly.

**Comment by glenn-jocher** [comment]
@isuchy it looks like you've done a thorough job converting your dataset and setting up the evaluation. The issue might be with the ONNX model inference or the dataset paths. Ensure your `data.yaml` paths are correct and that the ONNX model is loaded properly. You can also try running a few manual inferences to verify the model's outputs. If the problem persists, consider checking the ONNX model's compatibility with `onnxruntime` and verifying the bounding box coordinates. For further assistance, refer to the [Ultralytics documentation](https://docs.ultralytics.com/integrations/onnx/).
